File: java/sort_colors.java
public class Solution {
    void swap(int[] nums, int i, int j) {
        int tmp=nums[i];
        nums[i]=nums[j];
        nums[j]=tmp;
    }
    
    public void sortColors(int[] nums) {
        int n=nums.length;
        if (n==0) return;
        int start=0, end=n-1;
        for (int i=0; i<=end;) {
            if (nums[i]==0) {
                swap(nums, start, i);
                start++;
                i++;
            } else if (nums[i]==2) {
                swap(nums, end, i);
                end--;
            } else {
                i++;
            }
        }
    }
}

File: python/sort-colors.py
class Solution:
    def sortColors(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        n=len(nums)
        i, start, end=0, 0, n-1
        while i<=end:
            if nums[i]==0:
                if start!=i:
                    nums[start], nums[i]=nums[i], nums[start]
                start+=1
                i+=1
            elif nums[i]==1:
                i+=1
            else:
                nums[i], nums[end]=nums[end], nums[i]
                end-=1

